test/Basics/typeof.js

```diff
@@ -3,63 +3,98 @@
+            assert.areEqual("number", typeof 12345.678, "typeof 12345.678");
+            assert.areEqual("object", typeof {}, "typeof {}");
+            assert.areEqual("object", typeof arr, "typeof array");
+            assert.areEqual("number", typeof arr[0], "typeof arr[0]");
+            assert.areEqual("undefined", typeof arr[1], "typeof arr[1]");
+            assert.areEqual("object", typeof /abc/, "typeof /abc/");
+            assert.areEqual("function", typeof (function (){}), "typeof (function (){})");
+            assert.areEqual("function", typeof (() => 42), "typeof (() => 42)");
+            assert.areEqual("symbol", typeof Symbol(), "typeof Symbol()");
 
+            // built-in object and object wrappers
+            assert.areEqual("object", typeof (new String("")), "typeof empty string object wrapper");
+            assert.areEqual("object", typeof (new Boolean(false)), "typeof (new Boolean(false))");
+            assert.areEqual("object", typeof (new Number(0)), "typeof (new Number(0))");
+            assert.areEqual("object", typeof (new Number(12345.678)), "typeof (new Number(12345.678))");
+            assert.areEqual("object", typeof (new Object()), "typeof (new Object())");
+            assert.areEqual("object", typeof (new Array()), "typeof (new Array())");
+            assert.areEqual("object", typeof (new Date(123)), "typeof (new Date(123))");
+        }
+    },
+    {
+        name: "typeof of expressions",
+        body: function () {
+            function f() {
+                function g() { }
+                return g;
+            }
+            assert.areEqual("function", typeof f(), "typeof of function call");
+            assert.areEqual("number", typeof eval(7*6), "typeof of direct eval");
+        }
+    },
+    {
+        name: "typeof handling of undefined variables",
+        body: function () {
+            assert.areEqual("undefined", typeof x, "typeof of undeclaired var");
+            assert.areEqual("undefined", typeof {}.x, "typeof {}.x");
 
-WScript.Echo("typeof (new Number(12345.678)) : " + typeof (new Number(12345.678)));
-WScript.Echo("typeof (12345.678) : " + typeof (12345.678));
+            function test_var() { var res = typeof x; var x = 42; return res; }
+            assert.areEqual("undefined", test_var(), "typeof of hoisted var");
 
-function f() {
-    function g() { }
-    return g;
-}
+            function test_scoped() {
+                function inner() { var x = 42; } 
+                return typeof x;
+            }
+            assert.areEqual("undefined", test_scoped(), "typeof of var when it's out of scope");
 
-WScript.Echo("typeof function : " + typeof (f));
-WScript.Echo("typeof function returning function : " + typeof (f()));
+            function test_let() { var res = typeof x; let x = 42; return res; }
+            assert.throws(test_let, ReferenceError, "typeof of 'let' variable in its dead zone", "Use before declaration");

```

Comment:
Perhaps also add a case for `typeof x.y` when `x` is undefined

Reply:
this one throws (in V8 as well) -- expected?

---
In reply to: [205543506](https://github.com/Microsoft/ChakraCore/pull/5497#discussion_r205543506) [](ancestors = 205543506)

Reply:
Yes, it should throw.